File: games/friendzone4/src/main/java/fr/purpletear/friendzone4/game/activities/load/Load.java
package fr.purpletear.friendzone4.game.activities.load;

import android.content.Intent;
import android.os.Bundle;
import android.os.Parcelable;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;

import org.jetbrains.annotations.NotNull;

import fr.purpletear.friendzone4.R;
import fr.purpletear.friendzone4.game.config.ChapterDetailsHandler;
import fr.purpletear.friendzone4.game.config.Params;
import purpletear.fr.purpleteartools.TableOfSymbols;


public class Load extends AppCompatActivity {
    /**
     * Handles the model settings
     *
     * @see LoadModel
     */
    private LoadModel model;

    private boolean stop = false;
    private int hasSeenChapterNumber = -1;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.fz4_activity_load);

        load();
    }

    @Override
    protected void onSaveInstanceState(@NotNull Bundle outState) {
        super.onSaveInstanceState(outState);
        outState.putInt("hasSeenChapterNumber", this.hasSeenChapterNumber);
        outState.putParcelable("params", model.getParams());
        outState.putParcelable("symbols", model.getSymbols());
        outState.putBoolean("hasSeenAudioCinematic", model.hasSeenAudioCinematic);
        outState.putBoolean("hasSeenLoading", model.hasSeenLoading);
        outState.putBoolean("hasSeenTextCinematic", model.hasSeenTextCinematic);
        outState.putBoolean("hasSeenPoetry", model.hasSeenPoetry);
        outState.putBoolean("stop", stop);
    }

    @Override
    protected void onRestoreInstanceState(@NotNull Bundle savedInstanceState) {
        super.onRestoreInstanceState(savedInstanceState);
        this.hasSeenChapterNumber = savedInstanceState.getInt("hasSeenChapterNumber");
        model.hasSeenAudioCinematic = savedInstanceState.getBoolean("hasSeenAudioCinematic");
        model.hasSeenLoading = savedInstanceState.getBoolean("hasSeenLoading");
        model.hasSeenTextCinematic = savedInstanceState.getBoolean("hasSeenTextCinematic");
        model.hasSeenPoetry = savedInstanceState.getBoolean("hasSeenPoetry");
        stop = savedInstanceState.getBoolean("stop");
        model.setParams((Params) savedInstanceState.getParcelable("params"));
        model.setSymbols((TableOfSymbols) savedInstanceState.getParcelable("symbols"));
    }


    @Override
    protected void onResume() {
        super.onResume();
        if (stop) {
            stop = false;
            return;
        }
        this.hasSeenChapterNumber = model.getSymbols().getChapterNumber();
        stop = true;
        navigate();
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
        super.onActivityResult(requestCode, resultCode, data);

        if (resultCode == RESULT_CANCELED) {
            if (data != null && data.hasExtra("params")) {
                setResult(RESULT_OK, new Intent()
                        .putExtra("params", (Parcelable) data.getParcelableExtra("params"))
                        .putExtra("symbols", (Parcelable) data.getParcelableExtra("symbols"))
                );
            } else {
                setResult(RESULT_OK, new Intent()
                        .putExtra("params", (Parcelable) model.getParams())
                        .putExtra("symbols", (Parcelable) model.getSymbols())
                );
            }
            stop = true;
            finish();
            return;
        }
        if (requestCode == NavigationHandler.Navigation.TEXTCINEMATIC.ordinal()) {
            model.hasSeenTextCinematic = true;
            if (data != null && data.hasExtra("params")) {
                model.setParams((Params) data.getParcelableExtra("params"));
            }
            if (stop) {
                navigate();
            }
        } else if (requestCode == NavigationHandler.Navigation.AUDIOCINEMATIC.ordinal()) {
            model.hasSeenAudioCinematic = true;
            if (stop) {
                navigate();
            }
        } else if (requestCode == NavigationHandler.Navigation.LOADING.ordinal()) {
            model.hasSeenLoading = true;
            if (stop) {
                navigate();
            }
        } else if (requestCode == NavigationHandler.Navigation.GAME.ordinal()) {
            model.invalidate();
            if (data == null) {
                return;
            }
            model.setParams((Params) data.getParcelableExtra("params"));
            model.setSymbols((TableOfSymbols) data.getParcelableExtra("symbols"));
            if (stop) {
                navigate();
            }
        } else if (requestCode == NavigationHandler.Navigation.POETRY.ordinal()) {
            model.hasSeenPoetry = true;
            if (stop) {
                navigate();
            }
        }
    }

    /**
     * Loads the Activity's vars
     */
    private void load() {
        TableOfSymbols symbols = getIntent().getParcelableExtra("symbols");
        Params params = new Params();
        params.read(this);
        params.setChapterCode(symbols.getChapterCode());
        model = new LoadModel(
                params,
                symbols);
    }

    private void navigate() {
        model.getNavigationHandler().to(model.require());

        Intent i = model.getNavigationHandler().getIntent(this);

        i.putExtra("params", (Parcelable) model.getParams());
        i.putExtra("symbols", (Parcelable) model.getSymbols());
        i.putExtra("chapter", ChapterDetailsHandler.getChapter(this, model.getSymbols(), model.getParams().getChapterCode()));
        startActivityForResult(i.addFlags(Intent.FLAG_ACTIVITY_NO_ANIMATION), model.getNavigationHandler().getRequestCode());
        overridePendingTransition(0, 0);
    }

}

File: games/friendzone2/src/main/java/fr/purpletear/friendzone2/activities/load/Load.kt
/*
 * Copyright (C) PurpleTear, Inc - All Rights Reserved
 * Unauthorized copying of this file, via any medium is strictly prohibited
 * Proprietary and confidential
 * Written by Hocine Belbouab <hbdeveloppeur@gmail.com>, 1/10/19 11:38 AM
 *
 */

package fr.purpletear.friendzone2.activities.load

import android.content.Intent
import android.os.Bundle
import android.os.Parcelable
import androidx.appcompat.app.AppCompatActivity
import com.example.sharedelements.SutokoSharedElementsData
import fr.purpletear.friendzone2.R
import fr.purpletear.friendzone2.configs.ChapterDetailsHandler
import purpletear.fr.purpleteartools.GlobalData
import purpletear.fr.purpleteartools.TableOfSymbols

class Load : AppCompatActivity() {

    /**
     * Handles the model settings
     *
     * @see LoadModel
     */
    private lateinit var model: LoadModel

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_load)
        load()
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState.putParcelable("symbols", model.symbols)
        outState.putBoolean("hasSeenTextCinematic", model.hasSeenTextCinematic)
        outState.putBoolean("isGranted", model.isGranted)
    }

    override fun onRestoreInstanceState(savedInstanceState: Bundle) {
        super.onRestoreInstanceState(savedInstanceState)
        model.hasSeenTextCinematic = savedInstanceState.getBoolean("hasSeenTextCinematic")
        model.isGranted = savedInstanceState.getBoolean("isGranted")
        model.symbols = (savedInstanceState.getParcelable<Parcelable>("symbols") as TableOfSymbols)
    }

    override fun onResume() {
        super.onResume()
        navigate()
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)

        if (resultCode == RESULT_CANCELED) {
            finish()
            return
        }

        if (requestCode == NavigationHandler.Navigation.TEXTCINEMATIC.ordinal) {
            model.hasSeenTextCinematic = true
            if (data != null && data.hasExtra("symbols")) {
                model.symbols = (data.getParcelableExtra<Parcelable>("symbols") as TableOfSymbols)
            }
            if (data != null && data.hasExtra("soundPosition")) {
                model.soundPosition = data.getIntExtra("soundPosition", 0)
            }
        } else if (requestCode == NavigationHandler.Navigation.POETRY.ordinal) {
            model.hasSeenPoetry = true
        } else if (requestCode == NavigationHandler.Navigation.GAME.ordinal) {
            model.invalidate()
            if (data == null) {
                return
            }
            model.symbols = (data.getParcelableExtra<Parcelable>("symbols") as TableOfSymbols)
        }
    }

    /**
     * Loads the Activity's vars
     */
    private fun load() {
        val symbols =
            intent.getParcelableExtra("symbols") ?: TableOfSymbols(GlobalData.Game.FRIENDZONE2.id)

        if (SutokoSharedElementsData.SHOULD_FORCE_CHAPTER) {
            symbols.chapterCode = SutokoSharedElementsData.FORCE_CHAPTER_CODE
            symbols.save(this)
        }

        model = LoadModel(
            symbols, intent.getBooleanExtra("granted", false)
        )
    }

    @Throws
    private fun navigate() {

        model.navigationHandler.to(model.require())

        val i = model.navigationHandler.getIntent(this)

        if (i === null) {
            return
        }

        i.putExtra("symbols", model.symbols as Parcelable)
        i.putExtra(
            "chapter",
            ChapterDetailsHandler.getChapter(this, model.symbols.chapterCode, model.symbols)
        )

        startActivityForResult(
            i.addFlags(Intent.FLAG_ACTIVITY_NO_ANIMATION),
            model.navigationHandler.requestCode
        )
        overridePendingTransition(0, 0)
    }
}
